File: 006-ZigZagConversion/Python/main.py
class Solution:
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if not s:
            return ""
        elif numRows == 1:
            return s
        
        result = ""
        size = len(s)
        modulo = (2 * (numRows - 1))
        
        # initialization
        idx = [i for i in range(size) if i % modulo == 0]
        queue = [(idx[i], idx[i+1]) for i in range(len(idx) - 1)]
        
        # add last tuple if exists
        if idx[-1] < size:
            queue.append((idx[-1], idx[-1]+modulo))
        
        visited = {i: False for i in range(size)}
        
        while queue:
            new_queue = []
            beg, end = -1, -1
            for idx in queue:
                beg, end = idx
                if beg < size and not visited[beg]:
                    result += s[beg]
                    visited[beg] = True
                if end < size and not visited[end]:
                    result += s[end]
                    visited[end] = True
                if beg <= end:
                    new_queue.append((beg+1, end-1))
            
            queue, new_queue = new_queue, []
            
        return result
        

File: 006-ZigZagConversion/Go/main.go
package main

import "fmt"

type vrange struct {
    b int
    e  int
}

func convert(s string, numRows int) string {
    if s == "" {
        return ""
    } else if numRows == 1 {
        return s
    }
    
    result := ""
    size := len(s)
    modulo := 2 * (numRows - 1)
    
    // initialization
    var idx []int
    for i := 0; i < size; i++ {
        if i % modulo == 0 {
            idx = append(idx, i)
        }
    }
    
    var queue []vrange
    for i := 0; i < (len(idx) - 1); i++ {
        queue = append(queue, vrange{idx[i], idx[i+1]})
    }
    
    // add last tuple if exists
    if idx[len(idx) - 1] < size {
        queue = append(queue, vrange{idx[len(idx) - 1], idx[len(idx) - 1]+modulo})
    }
    
    visited := make(map[int]bool)
    for i := 0; i < size; i++ {
        visited[i] = false
    }
    
    for len(queue) != 0 {
        var new_queue []vrange
        beg, end := -1, -1
        for _, elem := range queue {
            beg, end = elem.b, elem.e
            if beg < size && !visited[beg] {
                result += string(s[beg])
                visited[beg] = true
            }
            if end < size && !visited[end] {
                result += string(s[end])
                visited[end] = true
            }
            if beg <= end {
                new_queue = append(new_queue, vrange{beg+1, end-1})
            }
        }
        queue = new_queue
    }

    return result
    
}

func main() {
    fmt.Println(convert("PAYPALISHIRING", 3))
}